src/planning/strategies/match-recipe-by-verb.ts

```diff
@@ -106,13 +109,20 @@ export class MatchRecipeByVerb extends Strategy {
               }
             }
 
-            function tryApplyHandleConstraint(name, connection, constraint, handle) {
-              if (connection.handle != null) {
-                return false;
-              }
-              if (!MatchRecipeByVerb.connectionMatchesConstraint(
-                      connection, constraint)) {
-                return false;
+            function tryApplyHandleConstraint(name: string, connSpec: ConnectionSpec, particle: Particle, connection: HandleConnection, constraint, handle) {
+              if (connection) {
+                if (connection.handle != null) {
+                  return false;
+                }
+                if (!MatchRecipeByVerb.connectionMatchesConstraint(connection, constraint)) {
+                  return false;
+                }
+              } else {
+                assert(connSpec);
+                if (!MatchRecipeByVerb.connectionMatchesConstraint(connSpec, constraint)) {
+                  return false;
+                }
```

Comment:
We could make this code a lot clearer:
```
if (connection && connection.handle != null) { // but with lazily created handles when would this be true?
  return false;
}
if (!MatchRecipeByVerb.connectionMatchesConstraint(connection || connSpec, constraint)) {
  return false;
}
connection = connection || particle.addConnectionName(connSpec.name);
```
Now it says:
(1) early-out if there's already a handle.
(2) check the connection matches the constraint.
(3) build the connection in the recipe if necessary.

There are a bunch more connectionMatchesConstraint calls below too, including some that you're adding. Can these be unified into the same logical flow?

Reply:
updated the code, like you suggested, thanks! 
similarly to the previous comment, i'd like to refrain from bigger refactorings in this class in this PR, if that's ok with you.

Reply:
In this case though you've added the extra paths here, so I think it's worth taking a brief closer look. If there's nothing obvious it's fine to pass though.